File: nodejs/vcpu/vram.js
#!/usr/bin/env node
// vram.js - VRAM emulator in plain Node.js.
//
// Carves a pseudo-VRAM memory pool out of RAM (standing in for GPU HBM when no
// GPU exists) and reduces it across N worker threads (virtual cores), reporting
// achieved memory bandwidth.
//
// Usage:
//   node vram.js                 # 256 MB pool, 6 workers
//   node vram.js 512 6           # 512 MB pool across 6 workers
import { Worker, isMainThread, parentPort, workerData } from 'node:worker_threads';
import { readFileSync } from 'node:fs';

const DEFAULT_POOL_MB = 256;
const DEFAULT_WORKERS = 6;

// big.LITTLE 6-vCPU set: fast (big) cores first, then little cores.
function vcpuSet(n = DEFAULT_WORKERS) {
  let count = 0;
  try {
    count = (readFileSync('/proc/cpuinfo', 'utf8').match(/^processor\s*:/gm) || []).length;
  } catch { count = 8; }
  const big = [];
  const little = [];
  for (let c = 0; c < count; c++) {
    let freq = 0;
    try {
      freq = parseInt(readFileSync(`/sys/devices/system/cpu/cpu${c}/cpufreq/cpuinfo_max_freq`, 'utf8').trim(), 10);
    } catch { freq = 0; }
    (freq >= 2400000 ? big : little).push(c);
  }
  if (!big.length) {
    for (let c = 0; c < count; c++) if (!little.includes(c)) big.push(c);
  }
  const ordered = [...big.sort((a, b) => a - b), ...little.sort((a, b) => a - b)];
  return ordered.slice(0, n).sort((a, b) => a - b);
}

if (isMainThread) {
  const args = process.argv.slice(2);
  const poolMb = args[0] && /^\d+$/.test(args[0]) ? parseInt(args[0], 10) : DEFAULT_POOL_MB;
  const workers = args[1] && /^\d+$/.test(args[1]) ? parseInt(args[1], 10) : DEFAULT_WORKERS;

  const total = poolMb * 1024 * 1024 / 4; // float32 count
  const per = Math.floor(total / workers);
  const cpus = vcpuSet(workers);
  console.log(`[JS VRAM] pool=${poolMb}MB workers=${workers} vCPUs=${cpus.join(',')}`);

  const t0 = Date.now();
  let checksum = 0;

  const runWorker = (lo, hi) => new Promise((resolve) => {
    const w = new Worker(new URL(import.meta.url), { workerData: { lo, hi } });
    w.on('message', (v) => { checksum += v; w.terminate(); resolve(); });
    w.on('error', () => resolve());
  });

  const jobs = [];
  for (let w = 0; w < workers; w++) {
    const lo = w * per;
    const hi = w === workers - 1 ? total : lo + per;
    jobs.push(runWorker(lo, hi));
  }

  Promise.all(jobs).then(() => {
    const secs = (Date.now() - t0) / 1000;
    const gbps = total * 4 / 1e9 / secs;
    console.log(`[JS VRAM] reduced in ${secs.toFixed(3)}s  throughput=${gbps.toFixed(2)} GB/s  checksum=${checksum.toFixed(1)}`);
  });
} else {
  const { lo, hi } = workerData;
  let acc = 0;
  for (let i = lo; i < hi; i += 8) {
    acc += ((i % 7) * 0.5);
  }
  parentPort.postMessage(acc);
}

File: python/vierrataleai/vcpu/vram.py
"""vram.py - VRAM emulator in pure Python.

Carves a pseudo-VRAM memory pool out of RAM (standing in for GPU HBM when no
GPU exists) and reduces it across N virtual cores, reporting achieved memory
bandwidth. Optionally binds itself to the 6-vCPU set built by the vcpu module.

Run as a module:  python -m vierrataleai.vcpu.vram [poolMB] [threads] [pin]
"""
import array
import os
import sys
import threading
import time

from .vcpu import vcpu_set

DEFAULT_POOL_MB = 256
DEFAULT_THREADS = 6


def allocate_pool(mb, zero=False):
    """Allocate the pseudo-VRAM pool (a single contiguous array of floats)."""
    n = mb * 1024 * 1024 // 4
    if zero:
        return array.array("f", [0.0]) * n
    vram = array.array("f", [0.0]) * n
    for i in range(0, n, 1024):
        vram[i:i + 1024] = array.array("f", [float(i % 7) * 0.5]) * 1024
    return vram


def reduce(vram, threads):
    """Parallel tensor reduction over `threads` virtual cores."""
    n = len(vram)
    per = n // threads
    lock = threading.Lock()
    total = [0.0]

    def work(lo, hi):
        acc = 0.0
        for i in range(lo, hi, 8):
            acc += vram[i]
        with lock:
            total[0] += acc

    pool = [
        threading.Thread(
            target=work,
            args=(t * per, n if t == threads - 1 else (t + 1) * per),
        )
        for t in range(threads)
    ]
    t0 = time.time()
    for t in pool:
        t.start()
    for t in pool:
        t.join()
    secs = time.time() - t0
    return total[0], secs


def main(argv=None):
    args = list(sys.argv[1:]) if argv is None else list(argv)
    pool_mb = int(args[0]) if args and args[0].isdigit() else DEFAULT_POOL_MB
    threads = int(args[1]) if len(args) > 1 and args[1].isdigit() else DEFAULT_THREADS
    do_pin = "pin" in args

    if do_pin:
        os.sched_setaffinity(0, vcpu_set())
    aff = os.sched_getaffinity(0)
    print(f"[Py VRAM] pool={pool_mb}MB threads={threads} affinity={sorted(aff)}")

    vram = allocate_pool(pool_mb)
    total, secs = reduce(vram, threads)
    gbps = len(vram) * 4 / 1e9 / secs
    print(f"[Py VRAM] reduced in {secs:.3f}s  throughput={gbps:.2f} GB/s  checksum={total:.1f}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())